test/inductor/test_torchinductor.py

```diff
@@ -8778,6 +8778,16 @@ def fn(x):
         x = torch.rand(128, 32, 63)
         self.common(fn, (x,))
 
+    def test_vectorized_ops_masked(self):
+        def fn(x):
+            index = torch.arange(64)
+            mask = index.view(1, 1, 64) < 63
+            indices = [None, None, index]
+            return torch.ops.aten._unsafe_masked_index(x, mask, indices, 7)
```

Comment:
This test doesn't generate `all_masked` from my side. Is it expected?
```c++
            #pragma omp for
            for(long x0=static_cast<long>(0L); x0<static_cast<long>(4096L); x0+=static_cast<long>(1L))
            {
                for(long x1=static_cast<long>(0L); x1<static_cast<long>(64L); x1+=static_cast<long>(16L))
                {
                    auto tmp0 = x1;
                    auto tmp1 = c10::convert<int32_t>(tmp0);
                    auto tmp2 = at::vec::Vectorized<int32_t>::arange(tmp1, 1);
                    auto tmp3 = static_cast<int32_t>(63);
                    auto tmp4 = at::vec::Vectorized<int32_t>(tmp3);
                    auto tmp5 = at::vec::VecMask<int32_t,1>(tmp2 < tmp4);
                    auto tmp6 = [&]
                    {
                        auto tmp7 = tmp5.template cast<float,1>().template loadu<float,1>(in_ptr0 + static_cast<long>(x1 + (63L*x0)));
                        return tmp7;
                    }
                    ;
                    auto tmp10 =
                    [&]
                    {
                        if (tmp5.all_zero())
                        {
                            return at::vec::Vectorized<float>(static_cast<float>(7));
                        }
                        else
                        {
                            auto tmp8 = tmp6();
                            auto tmp9 = at::vec::Vectorized<float>(static_cast<float>(7));
                            return decltype(tmp8)::blendv(tmp9, tmp8, tmp5.template cast<float,1>());
                        }
                    }
                    ()
                    ;
                    tmp10.store(out_ptr0 + static_cast<long>(x1 + (64L*x0)));
                }
            }
```

Reply:
Did you copy the test to python REPL? I get the same output as you if I do that. I get a different output when the test is run with a `TORCH_CHECK`

Reply:
Yes, can you paste the expected generated code (before and after) from your side?

Reply:
Main:

<details>

```python
# AOT ID: ['0_inference']
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty_strided
from torch._inductor.async_compile import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels
from torch._inductor.codegen.multi_kernel import MultiKernelCall

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
_quantized = torch.ops._quantized
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
empty_strided_cpu = torch._C._dynamo.guards._empty_strided_cpu
empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda
reinterpret_tensor = torch._C._dynamo.guards._reinterpret_tensor
alloc_from_pool = torch.ops.inductor._alloc_from_pool
async_compile = AsyncCompile()


cpp_fused__unsafe_masked_index_lt_0 = async_compile.cpp_pybinding(['const float*', 'float*'], '''
#include "/tmp/tmpfqa7he29/ky/cky2bufythacofebk7ujv36e4pxyqcqbpsy5r4vojoprjiwcwfxf.h"
extern "C"  void kernel(const float* in_ptr0,
                       float* out_ptr0)
{
    #pragma omp parallel num_threads(32)
    {
        int tid = omp_get_thread_num();
        {
            #pragma omp for
            for(long x0=static_cast<long>(0L); x0<static_cast<long>(4096L); x0+=static_cast<long>(1L))
            {
                for(long x1=static_cast<long>(0L); x1<static_cast<long>(64L); x1+=static_cast<long>(8L))
                {
                    auto tmp0 = x1;
                    auto tmp1 = c10::convert<int32_t>(tmp0);
                    auto tmp2 = at::vec::Vectorized<int32_t>::arange(tmp1, 1);
                    auto tmp3 = static_cast<int32_t>(63);
                    auto tmp4 = at::vec::Vectorized<int32_t>(tmp3);
                    auto tmp5 = at::vec::VecMask<int32_t,1>(tmp2 < tmp4);
                    auto tmp6 = [&]
                    {
                        auto tmp7 = c10::convert<int64_t>(tmp0);
                        auto tmp8 = at::vec::VectorizedN<int64_t,2>::arange(tmp7, 1);
                        TORCH_CHECK((at::vec::VecMask<int64_t,2>(tmp8 < at::vec::VectorizedN<int64_t,2>(63L))).all_masked(), "index out of bounds: tmp8 < 63L");
                        auto tmp10 = tmp5.template cast<float,1>().template loadu<float,1>(in_ptr0 + static_cast<long>(x1 + (63L*x0)));
                        return tmp10;
                    }
                    ;
                    auto tmp13 =
                    [&]
                    {
                        if (tmp5.all_zero())
                        {
                            return at::vec::Vectorized<float>(static_cast<float>(7));
                        }
                        else
                        {
                            auto tmp11 = tmp6();
                            auto tmp12 = at::vec::Vectorized<float>(static_cast<float>(7));
                            return decltype(tmp11)::blendv(tmp12, tmp11, tmp5.template cast<float,1>());
                        }
                    }
                    ()
                    ;
                    tmp13.store(out_ptr0 + static_cast<long>(x1 + (64L*x0)));
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (128, 32, 63), (2016, 63, 1))
    buf0 = empty_strided_cpu((128, 32, 64), (2048, 64, 1), torch.float32)
    cpp_fused__unsafe_masked_index_lt_0(arg0_1, buf0)
    del arg0_1
    return (buf0, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((128, 32, 63), (2016, 63, 1), device='cpu', dtype=torch.float32)
    fn = lambda: call([arg0_1])
    return print_performance(fn, times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```

</details>


PR:

<details>

```python
# AOT ID: ['0_inference']
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty_strided
from torch._inductor.async_compile import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels
from torch._inductor.codegen.multi_kernel import MultiKernelCall

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
_quantized = torch.ops._quantized
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
empty_strided_cpu = torch._C._dynamo.guards._empty_strided_cpu
empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda
reinterpret_tensor = torch._C._dynamo.guards._reinterpret_tensor
alloc_from_pool = torch.ops.inductor._alloc_from_pool
async_compile = AsyncCompile()


cpp_fused__unsafe_masked_index_lt_0 = async_compile.cpp_pybinding(['const float*', 'float*'], '''
#include "/tmp/tmpys5d1329/ky/cky2bufythacofebk7ujv36e4pxyqcqbpsy5r4vojoprjiwcwfxf.h"
extern "C"  void kernel(const float* in_ptr0,
                       float* out_ptr0)
{
    #pragma omp parallel num_threads(32)
    {
        int tid = omp_get_thread_num();
        {
            #pragma omp for
            for(long x0=static_cast<long>(0L); x0<static_cast<long>(4096L); x0+=static_cast<long>(1L))
            {
                for(long x1=static_cast<long>(0L); x1<static_cast<long>(64L); x1+=static_cast<long>(8L))
                {
                    auto tmp0 = x1;
                    auto tmp1 = c10::convert<int32_t>(tmp0);
                    auto tmp2 = at::vec::Vectorized<int32_t>::arange(tmp1, 1);
                    auto tmp3 = static_cast<int32_t>(63);
                    auto tmp4 = at::vec::Vectorized<int32_t>(tmp3);
                    auto tmp5 = at::vec::VecMask<int32_t,1>(tmp2 < tmp4);
                    auto tmp6 = [&]
                    {
                        auto tmp7 = c10::convert<int64_t>(tmp0);
                        auto tmp8 = at::vec::VectorizedN<int64_t,2>::arange(tmp7, 1);
                        TORCH_CHECK(((at::vec::VecMask<int64_t,2>(tmp8 < at::vec::VectorizedN<int64_t,2>(63L))) | ~(tmp5)).all_masked(), "index out of bounds: tmp8 < 63L");
                        auto tmp10 = tmp5.template cast<float,1>().template loadu<float,1>(in_ptr0 + static_cast<long>(x1 + (63L*x0)));
                        return tmp10;
                    }
                    ;
                    auto tmp13 =
                    [&]
                    {
                        if (tmp5.all_zero())
                        {
                            return at::vec::Vectorized<float>(static_cast<float>(7));
                        }
                        else
                        {
                            auto tmp11 = tmp6();
                            auto tmp12 = at::vec::Vectorized<float>(static_cast<float>(7));
                            return decltype(tmp11)::blendv(tmp12, tmp11, tmp5.template cast<float,1>());
                        }
                    }
                    ()
                    ;
                    tmp13.store(out_ptr0 + static_cast<long>(x1 + (64L*x0)));
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (128, 32, 63), (2016, 63, 1))
    buf0 = empty_strided_cpu((128, 32, 64), (2048, 64, 1), torch.float32)
    cpp_fused__unsafe_masked_index_lt_0(arg0_1, buf0)
    del arg0_1
    return (buf0, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((128, 32, 63), (2016, 63, 1), device='cpu', dtype=torch.float32)
    fn = lambda: call([arg0_1])
    return print_performance(fn, times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```

</details>

Reply:
Seems I cannot find `all_masked` in your pasted code. Did I miss anything?

Reply:
There's a line like
```
TORCH_CHECK(((at::vec::VecMask<int64_t,2>(tmp8 < at::vec::VectorizedN<int64_t,2>(63L))) | ~(tmp5)).all_masked(), "index out of bounds: tmp8 < 63L");
```

Reply:
Thanks. For this case, it seems the `TORCH_CHECK` is redundant. The code after it doesn't use `tmp8` at all and the masked load with `tmp5` already takes care of boundaries?

Reply:
Right. That's why I added the additional condition to TORCH_CHECK so that it passes instead of failing like on `main`.

Reply:
Note that `tmp8` is not used, but `tmp8` is `x1` in disguise, which is used in the indexing after the `TORCH_CHECK`.